ddd-koans/Value Object Pt. 2 & Entity/Note/src/Main.kt

```diff
@@ -0,0 +1,42 @@
+import NoteValue.*
+import java.math.BigDecimal
+import java.math.MathContext
+import java.time.Duration
+
+data class Note(val noteValue: NoteValue, val pitch: Pitch) : LengthInTempo by noteValue
+
+enum class NoteValue(private val relativeValue: Int): LengthInTempo {
+    WholeNote(1),
+    HalfNote(2),
+    QuarterNote(4),
+    EighthNote(8),
+    SixteenthNote(16),
+    ThirtySecondNote(32);
+
+    private val lengthInBeats: BigDecimal = 4.toBigDecimal().divide(relativeValue.toBigDecimal())
+
+    override fun lengthIn(tempo: Tempo): Length {
+        val minuteAsSeconds = Duration.ofMinutes(1L).toSeconds().toBigDecimal()
+        val durationInSeconds = minuteAsSeconds
+            .multiply(lengthInBeats).divide(tempo.bpm.toBigDecimal(), MathContext.DECIMAL32)
+        return Length(durationInSeconds)
+    }
+
+    operator fun invoke(pitchStringRepresentation: String) = Note(this, Pitch.valueOf(pitchStringRepresentation))
```

Comment:
so it can throw exception if provided `pitchStringRepresentation` doesn't correspond to `Pitch` enum, what is more it won't be quite `domain` related exception I guess - not sure if it's here we show best practisies

Reply:
You are totally right. So should I add InvalidPitchRepresentatonException?

Reply:
yup, it should be enough